### DISCUSSION TITLE
Transform CityGML to CityJSON into memory rather than file?

### DISCUSSION THREAD

**Discussion by SeppBerkner** [cityjson/specs#211] [Q&A] [answered]
Hello everyone 👋 

I am getting more and more versatile with the CItyJSON standard :) One thing that could improve the performance of my data pipeline is to transform the raw CityGML to CityJSON in memory rather than creating a new CityJSON file. Right now I have the following Python function:

```
def citygml_to_cityjson(target_dir: str, repo_root_dir: str):
    if not os.path.exists(target_dir):
        raise FileNotFoundError(f"Expected directory does not exist: {target_dir}")

    file_path = None
    for filename in os.listdir(target_dir):
        if filename.endswith(".gml"):
            file_path = os.path.join(target_dir, filename)

    if not file_path:
        raise FileNotFoundError("No CityGML found in the downloaded files.")

    file_name = os.path.basename(file_path)

    # convert downloaded CityGML into CityJSON using cli tool citygml-tools
    cli_command = [
        f"{repo_root_dir}/external_libs/citygml-tools/citygml-tools",
        "to-cityjson",
        file_path
    ]
    subprocess.run(cli_command, capture_output=False, text=False, check=False)
    
    for filename in os.listdir(target_dir):
        if filename.endswith("kachel.json"):
            file_path_json = os.path.join(target_dir, filename)
    
     with open(file_path_json, "r") as file:
         cityjson = json.load(file)
    
    return cityjson
```

I read in a CityGML file, convert it to a CityJSON file using the CLI command and load the content back in to pass it to the next steps of my data pipeline. I would love to skip the file creation and loading step. Is there any solution for this?

**Comment by hugoledoux** [comment]
this is a question for https://github.com/citygml4j/citygml-tools/ actually.

I don't think that it allows you to write to stdout out, I checked but don't see it. But @clausnagel knows better here.

**Comment by clausnagel** [comment, marked as answer]
That's correct, the current version of citygml-tools doesn't support writing to stdout. PRs are welcome :-)

But under Linux, simply use named pipes to avoid creating temporary files.

**Reply by SeppBerkner** [reply]
OK. Thanks for the tip with the named pipes. I will try that out
